### PR TITLE
Enable the deepseek OCR model on Gaudi.

### PR DESCRIPTION
**Pull request by HeJunyan**
The functionality is ready now, test by:

VLLM_SKIP_WARMUP=true python3 vllm-fork/examples/offline_inference/vision_language.py -m deepseek_ocr

TODO:
1. Padding and graph mode optimization for sam_model and vision_model inside it.
2. Full warmup support for multimodal input.

### CODE REVIEW

File: examples/offline_inference/vision_language.py [hunk 1]
```diff
@@ -150,6 +151,49 @@ def run_deepseek_vl2(questions: list[str], modality: str) -> ModelRequestData:
...
+    assert modality == "image"
+
+    model_name = "deepseek-ai/DeepSeek-OCR"
+
+    engine_args = EngineArgs(
+        model=model_name,
+        max_model_len=4096,
+        max_num_seqs=5,
+        limit_mm_per_prompt={modality: 1},
+#        enforce_eager=True,
```
Comment (Wei-Lin-Intel) [inline examples/offline_inference/vision_language.py]: delete it

File: vllm/model_executor/models/deepencoder.py [hunk 2]
```diff
@@ -0,0 +1,684 @@
...
+            )
+            rel_w = rel_w.view(
+                B, self.num_heads, rel_w.size(1), rel_w.size(2), rel_w.size(3)
+            )
+            attn_bias = (rel_h + rel_w).view(
+                B, self.num_heads, rel_h.size(2), rel_h.size(3) * rel_w.size(4)
+            )
+
+            x = FusedSDPA.apply(q, k, v, attn_bias, 0.0, False, None, self.softmax_mode)
+            #x = torch.nn.functional.scaled_dot_product_attention(
```
Comment (Wei-Lin-Intel) [inline vllm/model_executor/models/deepencoder.py]: delete these comments

File: vllm/model_executor/models/deepencoder.py [hunk 3]
```diff
@@ -0,0 +1,684 @@
...
+            embed_dim (int): Patch embedding dimension.
+        """
+        super().__init__()
+
+        self.proj = nn.Conv2d(
+            in_chans, embed_dim, kernel_size=kernel_size, stride=stride, padding=padding
+        )
+
+    def forward(self, x: torch.Tensor) -> torch.Tensor:
+        x = self.proj(x)
```
Comment (Wei-Lin-Intel) [inline vllm/model_executor/models/deepencoder.py:524]: Don't we need a BF16 cast for `x` here anymore?

File: vllm/model_executor/models/deepseek.py [hunk 4]
```diff
@@ -110,15 +111,23 @@ def __init__(
                 f"Tensor parallel size {self.tp_size} is greater than "
                 f"the number of experts {self.n_routed_experts}.")
 
-        self.experts = nn.ModuleList([
-            DeepseekMLP(hidden_size=config.hidden_size,
+#        self.experts = nn.ModuleList([
```
Comment (Wei-Lin-Intel) [inline vllm/model_executor/models/deepseek.py]: delete them

File: vllm/model_executor/models/deepseek_vl2.py [hunk 5]
```diff
@@ -104,22 +106,23 @@ def __init__(self, cfg: MlpProjectorConfig):
...
-        if pad > 0:
-            x = F.pad(x, (0, 0, 0, pad, 0, pad), "constant", 0)
-        """4 to 1 concat"""
-        x = x.permute(0, 3, 1, 2)  # B, C, H, W
-        x = F.unfold(x,
-                     kernel_size=self.cfg.downsample_ratio,
-                     stride=self.cfg.downsample_ratio,
-                     padding=0)  # B, C*4, HW // 4
-        x = x.permute(0, 2, 1)
+        if self.projector_type == "downsample_mlp_gelu":
```
Comment (Wei-Lin-Intel) [inline vllm/model_executor/models/deepseek_vl2.py:109]: Why we need a condition here?

Comment (HeJunyan) [reply vllm/model_executor/models/deepseek_vl2.py:109]: It is imported by upstream deepseek enabling patches.
image
And the upstream code is the same:
[https://github.com/vllm-project/vllm/blob/main/vllm/model_executor/models/deepseek_vl2.py#L135](url)

This deepseek model's projector_type name is not "downsample_mlp_gelu", so in order to bypass this calculation here while do not change the behavior of other models, the condition is added here, I think.

File: vllm/model_executor/models/deepencoder.py (hunk 3 above)
Comment (HeJunyan) [reply vllm/model_executor/models/deepencoder.py:524]: I change the input type when parsing and validating them:
image
So, we do not need to consider the type conversion in later stages.

I notice that the V1 engine of upstream will do the type conversion for your model automatically, and the input type is already correct. So in order to keep the same behavior with that, I choose to convert the type at the first stage. If you think that the conversion here has better performance, I can change to this way.

**Review by Wei-Lin-Intel — APPROVED** [review]
LGTM

File: vllm/model_executor/models/deepseek.py [hunk 6]
```diff
@@ -156,26 +156,21 @@ def pack_params(self):
         self.w2 = self.w2.view(len(w2), *w2s[0].shape)
 
     def forward(self, hidden_states: torch.Tensor) -> torch.Tensor:
-        num_tokens, hidden_dim = hidden_states.shape
```
Comment (czhu15) [inline vllm/model_executor/models/deepseek.py:159]: can you check if this new interface does not break other deepseek models?

**Review by czhu15 — COMMENTED** [review]
1, pls fix the pre-commit issues.
2, this PR introduced serval places that may impact common deepseek models. pls double check/test with other deepseek models to make sure it does not break them.
Thanks!

File: vllm/model_executor/models/deepseek.py (hunk 6 above)
Comment (Wei-Lin-Intel) [reply vllm/model_executor/models/deepseek.py:159]: This model is DeepSeek-V1, not v2/v3. Won't break other models.

**Comment by Wei-Lin-Intel** [comment]
@czhu15 Passed the pre-commit check
